test/replication-luatest/gh_6033_box_promote_demote_test.lua

```diff
@@ -0,0 +1,501 @@
+    wal_delay_start(g.server_1)
+
+    local wal_write_count = g.server_1:wal_write_count()
+    g.f = promote_start(g.server_1)
+    g.server_1:wait_wal_write_count(wal_write_count + 1)
+
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(not ok and err.code == box.error.UNSUPPORTED,
+        'error while promoting while in promote')
+
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.demote)
+    end)
+    luatest.assert(not ok and err.code == box.error.UNSUPPORTED,
+        'error while demoting while in promote')
+end
+
+g_nspaces.after_test('test_simultaneous', function(g)
+    -- Finish already started promote
+    wal_delay_end(g.server_1)
+    fiber_join(g.server_1, g.f)
+    g.f = nil
+    g.cluster:demote(g.server_1)
+    g.server_2:wait_synchro_queue_owner_id(0)
+end)
+
+-- Promoting voter should fail.
+g_nspaces.test_voter_promote = function(g)
+    g.server_1:box_config({election_mode = 'voter'})
+
+    local ok, err = g.server_1:exec(function()
+        return pcall(box.ctl.promote)
+    end)
+    luatest.assert(not ok and err.code == box.error.UNSUPPORTED,
+        'error while promoting voter')
+end
+
+-- Promoting should fail if it is interrupted from another server
+-- while writing wal.
+g_nspaces.test_wal_interfering_promote = function(g)
+    -- Promote server_2, while server_1 is stuck.
+    wal_delay_start(g.server_1)
+    local wal_write_count = g.server_1:wal_write_count()
+    g.server_2:promote()
+    g.server_1:wait_wal_write_count(wal_write_count + 2)
+
+    local election_term = g.server_1:election_term()
+    local f = promote_start(g.server_1)
+    g.server_1:wait_election_term(election_term + 1)
+
+    wal_delay_end(g.server_1)
+    local ok, err = fiber_join(g.server_1, f)
+
+    luatest.assert(not ok and err.code == box.error.INTERFERING_PROMOTE,
+        'interfering promote not handled')
+end
+
+g_nspaces.after_test('test_wal_interfering_promote', function(g)
+    g.cluster:promote(g.server_1)
+    g.cluster:demote(g.server_1)
+end)
+
+-- Demoting should fail if it is interrupted from another server
+-- while writing wal.
+g_nspaces.test_wal_interfering_demote = function(g)
+    g.cluster:promote(g.server_2)
+
+    -- Promote server_1, while server_2 is stuck.
+    wal_delay_start(g.server_2)
+    local wal_write_count = g.server_2:wal_write_count()
+    g.server_1:promote()
+    g.server_2:wait_wal_write_count(wal_write_count + 2)
+
+    local election_term = g.server_2:election_term()
+    local f = demote_start(g.server_2)
+    g.server_2:wait_election_term(election_term + 1)
+
+    wal_delay_end(g.server_2)
+    local ok, err = fiber_join(g.server_2, f)
+
+    luatest.assert(not ok and err.code == box.error.INTERFERING_PROMOTE,
+        'interfering demote not handled')
+end
+
+g_nspaces.after_test('test_wal_interfering_demote', function(g)
+    g.cluster:promote(g.server_2)
```

Comment:
Same question as in the previous test.
Here we know that server_2 has failed demotion because it received server1 promote first.
So let's simply demote it now.

Reply:
Same here. But now server_1 has limbo, but isn't leader in new term (server_2 demote one)